src/reducers/vis-state-merger.js

```diff
@@ -52,43 +51,49 @@ export function mergeFilters(state, filtersToMerge) {
 
   // merge filters
   filtersToMerge.forEach(filter => {
-    // match filter.dataId with current datesets id
-    // uploaded data need to have the same dataId with the filter
-    if (datasets[filter.dataId]) {
-      // datasets is already loaded
-      const validateFilter = validateFilterWithData(
-        datasets[filter.dataId],
-        filter
-      );
+    const datasetIds = Object.keys(datasets).reduce((acc, datasetId) => ([
+      ...acc,
+      ...(shouldApplyfilter(filter, datasetId) ? [datasetId] : [])
+    ]), []);
+
+    if (datasetIds.length) {
+
+      // for each dataset data mapped against the current filter are valid
+      const validatedFilter = datasetIds.reduce((acc, d) => {
+        const dataset = datasets[d];
+        const updatedFilter = validateFilterWithData(dataset, filter);
+        if (updatedFilter) {
+          return acc ? {
+            ...acc,
+            // merge props from every single dataset/field
+            ...mergeFilterProps(acc, updatedFilter)
+          } : updatedFilter;
+        }
+      }, null);
 
-      if (validateFilter) {
-        merged.push(validateFilter);
+      if (validatedFilter) {
+        merged.push(validatedFilter);
       }
     } else {
-      // datasets not yet loaded
       unmerged.push(filter);
     }
   });
 
   // filter data
   const updatedFilters = [...(state.filters || []), ...merged];
-  const datasetToFilter = uniq(merged.map(d => d.dataId));
-
-  const updatedDataset = datasetToFilter.reduce(
-    (accu, dataId) => ({
-      ...accu,
-      [dataId]: {
-        ...datasets[dataId],
-        ...filterData(datasets[dataId].allData, dataId, updatedFilters)
-      }
-    }),
-    datasets
-  );
+
+  // flatten all filter dataIds
+  const datasetsToFilter = uniq(merged.reduce((acc, f) => ([
```

Comment:
A more readabble version of this would be

`
import flattenDeep from 'lodash.flattendeep`
const datasetsToFilter = uniq(flattenDeep(merged.map(f => f.dataId)))
`